README.md
[![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg?style=flat-square)](http://makeapullrequest.com)
[![PowerShell Gallery](https://img.shields.io/powershellgallery/v/PSEmailRep.svg)](https://www.powershellgallery.com/packages/PSEmailRep)
[![PowerShell Gallery](https://img.shields.io/powershellgallery/dt/PSEmailRep.svg)](https://www.powershellgallery.com/packages/PSEmailRep)

# PSEmailRep

PSEmailRep is a simple PowerShell module to interface with the EmailRep.io API.

![ScreenShot](/Media/screenshot.png)

## What is EmailRep.io?

[EmailRep.io](https://emailrep.io/) EmailRep is a system of crawlers, scanners and enrichment services that collects data on email addresses, domains, and internet personas.

EmailRep uses hundreds of data points from social media profiles, professional networking sites, dark web credential leaks, data breaches, phishing kits, phishing emails, spam lists, open mail relays, domain age and reputation, deliverability, and more to predict the risk of an email address.

# PSEmailRep Commands

## Get-EmailRep

```powershell
> Get-EmailRep -EmailAddress bill@microsoft.com

email                      : bill@microsoft.com
reputation                 : high
suspicious                 : False
references                 : 79
blacklisted                : False
malicious_activity         : False
malicious_activity_recent  : False
credentials_leaked         : True
credentials_leaked_recent  : False
data_breach                : True
first_seen                 : 07/01/2008
last_seen                  : 05/24/2019
domain_exists              : True
domain_reputation          : high
new_domain                 : False
days_since_domain_creation : 10354
suspicious_tld             : False
spam                       : False
free_provider              : False
disposable                 : False
deliverable                : True
accept_all                 : True
valid_mx                   : True
spoofable                  : False
spf_strict                 : True
dmarc_enforced             : True
profiles                   : {flickr, myspace, spotify, pinterest...}
```

## New-EmailRep
```powershell
> New-EmailRep -Email test_emailrep_xxxxx@foobar.com -Tags bec -Description 'Business email takeover' -ApiKey $ApiKey

Status
------
Success
```

## Set-EmailRep
```powershell
> Set-EmailRep -APIKey

APIKey: ********************************
```

## Additional Documentation
[Additional Documentation](/Docs)

# Installing PSEmail

## PSGallery
```powershell
Install-Module -Name PSEmailRep -Repository PSGallery

Import-Module -Name PSEmailRep
```

## Clone the repo
```powershell
git clone https://github.com/arnydo/psemailrep.git

cd psemailrep/psemailrep

Import-Module PSEmailRep.psd1
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.2.2] - 2020-10-13
### Updated
- Parameter descriptions
- Minor tweaks

## [1.2.1] - 2020-02-09

### Updated
- UserAgent to reflect proper [RFC7231](https://tools.ietf.org/html/rfc7231#section-5.5.3) UA naming ("PSEmailRep/\*.\*.\*")


## [1.2.0] - 2020-02-07

### Added
- API authentication for `Get-EmailRep`
- `APIStatus` parameter to return API Quota status (Daily and/or Monthly)
- CHANGELOG.md to track changes to project

### Updated
- Copyright dates
- Module documentation files

tests/Help.Tests.ps1
$script:ModuleName = 'PSEmailRep'

# Removes all versions of the module from the session before importing
Get-Module $ModuleName | Remove-Module

$ModuleBase = Split-Path -Parent $MyInvocation.MyCommand.Path

# Get the list of functions we are not going to run tests against
$FunctionHelpTestExceptions = Get-Content -Path "$ModuleBase\Help.Exceptions.txt"

# For tests in .\Tests subdirectory
if ((Split-Path $ModuleBase -Leaf) -eq 'Tests') {
    $ModuleBase = Split-Path $ModuleBase -Parent | Join-Path -ChildPath 'PSEmailRep'
}

$Module = Import-Module $ModuleBase\$ModuleName.psd1 -PassThru -ErrorAction Stop
$commands = Get-Command -Module $module -CommandType Cmdlet, Function, Workflow  # Not alias

## When testing help, remember that help is cached at the beginning of each session.
## To test, restart session.

foreach ($command in $commands) {
    $commandName = $command.Name

    # Skip all functions that are on the exclusions list
    if ($script:FunctionHelpTestExceptions -contains $commandName) { continue } ## may not be correct check with a function that needs exceptions

    # The module-qualified command fails on Microsoft.PowerShell.Archive cmdlets
    $Help = Get-Help $commandName -ErrorAction SilentlyContinue

    Describe "Test help for $commandName" -Tag Help {

        # If help is not found, synopsis in auto-generated help is the syntax diagram
        It "should not be auto-generated" {
            $Help.Synopsis | Should Not BeLike '*`[`<CommonParameters`>`]*'
        }

        # Should be a description for every function
        It "gets description for $commandName" {
            $Help.Description | Should Not BeNullOrEmpty
        }

        # Should be at least one example
        It "gets example code from $commandName" {
            ($Help.Examples.Example | Select-Object -First 1).Code | Should Not BeNullOrEmpty
        }

        # Should be at least one example description
        It "gets example help from $commandName" {
            ($Help.Examples.Example.Remarks | Select-Object -First 1).Text | Should Not BeNullOrEmpty
        }

        Context "Test parameter help for $commandName" {

            $Common = 'Debug', 'ErrorAction', 'ErrorVariable', 'InformationAction', 'InformationVariable', 'OutBuffer', 'OutVariable',
            'PipelineVariable', 'Verbose', 'WarningAction', 'WarningVariable'

            $parameters = $command.ParameterSets.Parameters | Sort-Object -Property Name -Unique | Where-Object Name -notin $common
            $parameterNames = $parameters.Name
            $HelpParameterNames = $Help.Parameters.Parameter.Name | Sort-Object -Unique

            foreach ($parameter in $parameters) {
                $parameterName = $parameter.Name
                $parameterHelp = $Help.parameters.parameter | Where-Object Name -EQ $parameterName

                # Should be a description for every parameter
                It "gets help for parameter: $parameterName : in $commandName" {
                    $parameterHelp.Description.Text | Should Not BeNullOrEmpty
                }

                # Required value in Help should match IsMandatory property of parameter
                It "help for $parameterName parameter in $commandName has correct Mandatory value" {
                    $codeMandatory = $parameter.IsMandatory.toString()
                    $parameterHelp.Required | Should Be $codeMandatory
                }

                # Parameter type in Help should match code
                It "help for $commandName has correct parameter type for $parameterName" {
                    $codeType = $parameter.ParameterType.Name
                    # To avoid calling Trim method on a null object.
                    $helpType = if ($parameterHelp.parameterValue) { $parameterHelp.parameterValue.Trim() }
                    $helpType | Should be $codeType
                }
            }

            foreach ($helpParm in $HelpParameterNames) {
                # Shouldn't find extra parameters in help.
                It "finds help parameter in code: $helpParm" {
                    $helpParm -in $parameterNames | Should Be $true
                }
            }
        }
    }
}


tests/Feature.Tests.ps1
$script:ModuleName = 'PSEmailRep'

# Removes all versions of the module from the session before importing
Get-Module $ModuleName | Remove-Module

$ModuleBase = Split-Path -Parent $MyInvocation.MyCommand.Path

# For tests in .\Tests subdirectory
if ((Split-Path $ModuleBase -Leaf) -eq 'Tests') {
    $ModuleBase = Split-Path $ModuleBase -Parent | Join-Path -ChildPath 'PSEmailRep'
}

## This variable is for the VSTS tasks and is to be used for referencing any mock artifacts
$Env:ModuleBase = $ModuleBase

Import-Module $ModuleBase\$ModuleName.psd1 -PassThru -ErrorAction Stop | Out-Null
Describe "Basic function feature tests" -Tags Build {

}


tests/Project.Tests.ps1
if (-not(Get-Module -ListAvailable -Name "PSScriptAnalyzer")) {
    Write-Warning "Installing latest version of PSScriptAnalyzer"
    # Install PSScriptAnalyzer
    Install-Module PSScriptAnalyzer -Force -Scope CurrentUser
}

$script:ModuleName = 'PSEmailRep'

$ModuleBase = Split-Path -Parent $MyInvocation.MyCommand.Path

# Get the list of Pester Tests we are going to skip
$PesterTestExceptions = Get-Content -Path "$ModuleBase\Project.Exceptions.txt"

# For tests in .\Tests subdirectory
if ((Split-Path $ModuleBase -Leaf) -eq 'Tests') {
    $ModuleBase = Split-Path $ModuleBase -Parent | Join-Path -ChildPath 'PSEmailRep'
}

Describe "PSScriptAnalyzer rule-sets" -Tag Build , ScriptAnalyzer {

    $Rules = Get-ScriptAnalyzerRule
    $scripts = Get-ChildItem $ModuleBase -Include *.ps1, *.psm1, *.psd1 -Recurse | Where-Object fullname -notmatch 'classes'

    foreach ( $Script in $scripts )
    {
        Context "Script '$($script.FullName)'" {

            foreach ( $rule in $rules )
            {
                # Skip all rules that are on the exclusions list
                if ($PesterTestExceptions -contains $rule.RuleName) { continue }
                It "Rule [$rule]" {

                    (Invoke-ScriptAnalyzer -Path $script.FullName -IncludeRule $rule.RuleName ).Count | Should Be 0
                }
            }
        }
    }
}


Describe "General project validation: $moduleName" -Tags Build {
    BeforeAll {
        Get-Module $ModuleName | Remove-Module
    }
    It "Module '$moduleName' can import cleanly" {
        {Import-Module $ModuleBase\$ModuleName.psd1 -force } | Should Not Throw
    }
}


build.ps1
param(
    [string[]]$Tasks
)

function Install-Dependency([string] $Name)
{
    $policy = Get-PSRepository -Name "PSGallery" | Select-Object -ExpandProperty "InstallationPolicy"
    Write-Output $Policy
    if($policy -ne "Trusted") {
        Set-PSRepository -Name "PSGallery" -InstallationPolicy Trusted
    }

    if (!(Get-Module -Name $Name -ListAvailable)) {
        Install-Module -Name $Name -Scope CurrentUser -Confirm:$False -Force
    }
}

function Analyze-Scripts
{
    param(
        [string]$Path = "$PSScriptRoot\PSEmailRep"
    )
    $result = Invoke-ScriptAnalyzer -Path $Path -Severity @('Error', 'Warning') -Recurse
    if ($result) {
       $result | Format-Table
       Write-Error -Message "$($result.SuggestedCorrections.Count) linting errors or warnings were found. The build cannot continue."
       EXIT 1
    }
}

function Run-Tests
{
    param(
        [string]$Path = "$PSScriptRoot\tests"
    )

    $results = Invoke-Pester -Path $Path -CodeCoverage $Path\*\*\*\*.ps1 -PassThru -Quiet
    if($results.FailedCount -gt 0) {
       Write-Output "  > $($results.FailedCount) tests failed. The build cannot continue."
       foreach($result in $($results.TestResult | Where-Object {$_.Passed -eq $false} | Select-Object -Property Describe,Context,Name,Passed,Time)){
            Write-Output "    > $result"
       }

       EXIT 1
    }
    $coverage = [math]::Round($(100 - (($results.CodeCoverage.NumberOfCommandsMissed / $results.CodeCoverage.NumberOfCommandsAnalyzed) * 100)), 2);
    Write-Output "  > Code Coverage: $coverage%"
}

function Deploy-Modules
{
    param(
        [string]$Path = "$PSScriptRoot"
    )

    try {

       foreach($module in $(Get-ChildItem -Path $Path | Where-Object {$_.Name.ToLower().Contains("psdeploy")})){
         $name = $module.Name.Split(".")[0]
         $localManifest = Import-PowerShellDataFile -Path $(Join-Path -Path $module.DirectoryName -ChildPath "$name\$name.psd1")
         if([Version]$localManifest.ModuleVersion -gt $(Find-Module -Name $name -Repository "PowerShell" -ErrorAction SilentlyContinue).Version) {
             Write-Output "> Deploying $name"
             Invoke-PSDeploy -Path $module.FullName -Force
         } else {
             Write-Output "Latest version already deployed."
         }
       }
    }
    catch {
       Write-Output $_.Exception.Message
       EXIT 1
    }
}

foreach($task in $Tasks){
    switch($task)
    {
        "analyze" {
            Install-Dependency -Name "PSScriptAnalyzer"
            Write-Output "Analyzing Scripts..."
            Analyze-Scripts
        }
        "test" {
            Install-Dependency -Name "Pester"
            Write-Output "Running Pester Tests..."
            Run-Tests
        }
        "deploy" {
            Install-Dependency -Name "PSDeploy"
            Write-Output "Deploying Modules..."
            Deploy-Modules
        }
    }
}



PSEmailRep/Public/New-EmailRep.ps1
function New-EmailRep {
    <#
    .SYNOPSIS
    Submit a new report to EmailRep.io
    
    .DESCRIPTION
    Reports an email address. Date of malicious activity defaults to the current time unless otherwise specified.
    
    .PARAMETER Email
    Email address being reported.
    
    .PARAMETER Tags
    Tags that should be applied.

    account_takeover - Legitimate email has been taken over by a malicious actor
    bec - Business email compromise, whaling, contact impersonation/display name spoofing
    brand_impersonation - Impersonating a well-known brand (e.g. Paypal, Microsoft, Google, etc.)
    browser_exploit - The hosted website serves an exploit
    credential_phishing - Attempting to steal user credentials
    generic_phishing - Generic phishing, should only be used if others don’t apply or a more specific determination can’t be made or would be too difficult
    malware - Malicious documents and droppers. Can be direct attachments, indirect free file hosting sites or droppers from malicious websites
    scam - Catch-all for scams. Sextortion, payment scams, lottery scams, investment scams, fake bank scams, etc.
    spam - Unsolicited spam or spammy behavior (e.g. forum submissions, unwanted bulk email)
    spoofed - Forged sender email (e.g. the envelope from is different than the header from)
    task_request - Request that the recipient perform a task (e.g. gift card purchase, update payroll, send w-2s, etc.)
    threat_actor - Threat actor/owner of phishing kit

    
    .PARAMETER Description
    Additional information and context.
    
    .PARAMETER Timestamp
    When this activity occurred in UTC. Defaults to now().
    
    .PARAMETER Expires
    Number of hours the email should be considered risky (suspicious=true and blacklisted=true in the QueryResponse). Defaults to no expiration unless account_takeover tag is specified, in which case the default is 14 days.
    
    .PARAMETER ApiKey
    API key granting access to EmailRep API.
    
    .PARAMETER UserAgent
    Specify the user agent of the web request. Defaults to "PSEmailRep/*.*.*"

    .PARAMETER Force
    Submit report without confirmation prompt.

    .EXAMPLE
    New-EmailRep -Email test_emailrep_xxxxx@foobar.com -Description "Business email compromise" -Tags bec -Force

    status 
    ------ 
    success
    
    .NOTES
    https://emailrep.io/docs/#report-an-email-address
    #>
     

    [Cmdletbinding(SupportsShouldProcess = $true, ConfirmImpact = 'High')]
    
    param(
        [Parameter(Mandatory = $true,
            Position = 0,
            ValueFromPipeline = $true,
            ValueFromPipelineByPropertyName = $true,
            HelpMessage = 'Email address to query'
        )]
        [ValidatePattern('^([\w-\.]+)@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.)|(([\w-]+\.)+))([a-zA-Z]{2,4}|[0-9]{1,3})(\]?)$')]
        [Alias('EmailAddress')]
        [string[]]
        $Email,

        [Parameter(
            Mandatory = $true,
            HelpMessage = 'Tags that should be applied'
        )]
        [ValidateSet(
            'account_takeover',
            'bec',
            'brand_impersonation',
            'browser_exploit',
            'credential_phishing',
            'generic_phishing',
            'malware',
            'scam',
            'spam',
            'spoofed',
            'task_request',
            'threat_actor'
        )]
        [string[]]
        $Tags,

        [Parameter(
            HelpMessage = 'Additional information and context'
        )]
        [string]
        $Description,

        [Parameter(
            HelpMessage = 'When this activity occurred in UTC. Defaults to now().'
        )]
        [int]
        $Timestamp = [int][double]::Parse((Get-Date -UFormat %s)),

        [Parameter(
            HelpMessage = 'Number of hours the email should be considered risky (suspicious=true and blacklisted=true in the QueryResponse). Defaults to no expiration unless account_takeover tag is specified, in which case the default is 14 days'
        )]
        [int]
        $Expires = 14,

        [Parameter(
            Mandatory = $false
        )]
        [string]
        $ApiKey,

        [ValidatePattern('\w')]
        [string]$UserAgent = $Script:DefaultUA,

        [Parameter(
            HelpMessage = 'Submit report without confirming'
        )]
        [switch]
        $Force

        
    )

    begin {

        if (-not $PSBoundParameters.ContainsKey('Confirm')) {
            $ConfirmPreference = $PSCmdlet.SessionState.PSVariable.GetValue('ConfirmPreference')
        }
        if (-not $PSBoundParameters.ContainsKey('WhatIf')) {
            $WhatIfPreference = $PSCmdlet.SessionState.PSVariable.GetValue('WhatIfPreference')
        }

        if (-not $ApiKey) {

            # Try and retrieve API Key stored in AppData
            try {
                Write-Verbose "Attempting to retrieve API Key from encrypted file."
                
                $APIKey = Get-EmailRepAPIKey

                Write-Verbose "API Key found!"

                if ($null -eq $ApiKey) {
                    Write-Warning "API Key is not set. Please set APIKey parameter or use 'Set-EmailRep -APIKey'"
                    throw "API Key not set"
                }
            }
            catch {
                Write-Warning "API Key not set and/or could not be retrieved. Please try again..."
                break
            }
        }
        
        $baseUrl = "https://emailrep.io/report"

        $headers = @{
            "Content-Type" = "application/json"
            'Key'          = $ApiKey
        }
    }

    process {

        foreach ($EmailAddress in $Email) {

            try {

                $Json = [PSCustomObject]@{
                    email       = $EmailAddress
                    tags        = $tags
                    description = $Description
                    timestamp   = $Timestamp
                    expires     = $Expires
                } | convertto-Json

                if ($force -or $PSCmdlet.ShouldProcess($json , "Reporting to EMailRep.io")) {
                    $response = Invoke-WebRequest -Method POST -Uri $baseUrl -Headers $headers -Body $json -UserAgent $UserAgent
                    $response.Content | ConvertFrom-JSON
                }
            }
            catch {
                $errorDetails = $null
                $response = $_.Exception | Select-Object -ExpandProperty 'message' -ErrorAction Ignore
                if ($response) {
                    $errorDetails = $_.ErrorDetails
                }
                
                if ($null -eq $errorDetails) {
                    Switch ($response) {
                        'The remote server returned an error: (400) Bad Request.' {
                            Write-Error -Message 'Bad Request - the account does not comply with an acceptable format.'
                        }
                        # Windows PowerShell 401 response
                        'The remote server returned an error: (401) Unauthorized.' {
                            Write-Error -Message 'Response status code does not indicate success: 401 (Unauthorized).'
                        }
                        # PowerShell Core 401 response
                        'Response status code does not indicate success: 401 (Unauthorized).' {
                            Write-Error -Message 'Response status code does not indicate success: 401 (Unauthorized).'
                        }
                        'The remote server returned an error: (403) Forbidden.' {
                            Write-Error -Message 'Forbidden - no user agent has been specified in the request.'
                        }
                        # Windows PowerSHell 404 response
                        'The remote server returned an error: (404) Not Found.' {
                            Write-Error -Message 'Path not found.'
                        }
                        # PowerShell Core 404 response
                        'Response status code does not indicate success: 404 (Not Found).' {
                            Write-Error -Message 'Path not found.'
                        }
                    }
                }
                else {
                    Write-error -Message ('Request to "{0}" failed: {1}' -f $baseUrl, $errorDetails)
                }
            }

        }

    }
}

PSEmailRep/Public/Set-EmailRep.ps1
function Set-EmailRep {
    <#
    .SYNOPSIS
    Set API Key for persistent use.
    
    .DESCRIPTION
    Set API Key for persistent use in encrypted string.
    Default file location is in user's AppData folder.
    
    .PARAMETER APIKey
    Set new API Key.
    
    .PARAMETER Test
    Test retrieval of API Key.
    
    .EXAMPLE
    Set-EmailRep -API

    API Key: ***************************************
    
    #>
    

    [cmdletbinding()]
    
    param(
        [switch]$APIKey,

        [switch]$Test
    )

    if ($APIKey) {
        try {
            if (-not (Test-Path $env:appdata\PSEmailRep\api.txt)) {
                New-Item -Path $env:appdata -Name PSEmailRep -ItemType Directory -Force | Out-Null
            }
            else {
                Write-Warning "API Key already configured. Overwrite existing?"
                $response = Read-Host -Prompt "Y/N"

                switch ($response) {
                    "Y" { Write-Host "Overwriting existing key..." -ForegroundColor Yellow }
                    "N" {
                        Write-Host "Keeping exising. Exiting..." -ForegroundColor Yellow
                        return
                    }
                    default {
                        Write-Host "Response unknown. Try again..."
                        return
                    }
                } 

            }
    
            $Key = Read-Host -Prompt "API Key" -AsSecureString



            $Key | ConvertFrom-SecureString | Out-File $env:appdata\PSEmailRep\api.txt
        }
        catch {

        }

    }

    if ($test) {
        $testkey = Get-Content $env:appdata\PSEmailRep\api.txt | ConvertTo-SecureString
        $Credential = New-Object -TypeName System.Management.Automation.PSCredential -ArgumentList "API Key", $testkey
        $Credential.GetNetworkCredential().password
    }

    
}

PSEmailRep/Public/Get-EmailRep.ps1
function Get-EmailRep {
    <#
    .SYNOPSIS
    Query the EmailRep.io API for a report on an email address.
    
    .DESCRIPTION
    EmailRep uses hundreds of data points from social media profiles, professional networking sites,
    dark web credential leaks, data breaches, phishing kits, phishing emails, spam lists,
    open mail relays, domain age and reputation, deliverability,
    and more to predict the risk of an email address. https://emailrep.io/docs/#emailrep-alpha-api
    
    .PARAMETER EmailAddress
    Email address to be queried.
    
    .PARAMETER Summary
    When set to $true, a summary about the email address will be returned.

    .PARAMETER APIKey
    API key to authenticate against api.
    
    .PARAMETER Raw
    The original API response will be displayed. By default, the 'details' object is not a nested object.

    .PARAMETER UserAgent
    Specify the user agent of the web request. Defaults to "PSEmailRep/*.*.*"

    .PARAMETER APIStatus
    Return current query quota status. Daily or Monthly based on type of API key being used.
    
    .EXAMPLE
    Get-EmailRep -EmailAdress bill@microsoft.com

    email                      : bill@microsoft.com
    reputation                 : high
    suspicious                 : False
    references                 : 79
    blacklisted                : False
    malicious_activity         : False
    malicious_activity_recent  : False
    credentials_leaked         : True
    credentials_leaked_recent  : False
    data_breach                : True
    first_seen                 : 07/01/2008
    last_seen                  : 05/24/2019
    domain_exists              : True
    domain_reputation          : high
    new_domain                 : False
    days_since_domain_creation : 10354
    suspicious_tld             : False
    spam                       : False
    free_provider              : False
    disposable                 : False
    deliverable                : True
    accept_all                 : True
    valid_mx                   : True
    spoofable                  : False
    spf_strict                 : True
    dmarc_enforced             : True
    profiles                   : {twitter, vimeo, angellist, linkedin...}

    .EXAMPLE
    Get-EmailRep -EmailAdress bill@microsoft.com -Raw

    email      : bill@microsoft.com
    reputation : high
    suspicious : False
    references : 79
    details    : @{blacklisted=False; malicious_activity=False; malicious_activity_recent=False; credentials_leaked=True; credentials_leaked_recent=False; 
                data_breach=True; first_seen=07/01/2008; last_seen=05/24/2019; domain_exists=True; domain_reputation=high; new_domain=False;
                days_since_domain_creation=10354; suspicious_tld=False; spam=False; free_provider=False; disposable=False; deliverable=True; accept_all=True;     
                valid_mx=True; spoofable=False; spf_strict=True; dmarc_enforced=True; profiles=System.Object[]}

    .EXAMPLE
    Get-EmailRep -EmailAdress bill@microsoft.com -APIStatus

    Daily queries remaining: 92
    ...
    
    .EXAMPLE
    "bill@microsoft.com" | Get-EmailRep

    .EXAMPLE
    "bill@microsoft.com","john@microsoft.com" | Get-EmailRep | ft email,reputation,credentials_leaked

    email              reputation credentials_leaked
    -----              ---------- ------------------
    bill@microsoft.com high                     True
    john@microsoft.com high                     True
    
    .NOTES
    https://emailrep.io/docs/#emailrep-alpha-api
    #>
    

    [Cmdletbinding()]
    
    param(
        [Parameter(Mandatory = $true,
            Position = 0,
            ValueFromPipeline = $true,
            ValueFromPipelineByPropertyName = $true,
            HelpMessage = 'Email address to query'
        )]
        [ValidatePattern('^([\w-\.]+)@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.)|(([\w-]+\.)+))([a-zA-Z]{2,4}|[0-9]{1,3})(\]?)$')]
        [string[]]
        $EmailAddress,

        [Parameter(
            HelpMessage = 'Return a summary property in the response'
        )]
        [switch]
        $Summary,

        [Parameter(
            Mandatory = $false
        )]
        [string]
        $ApiKey,

        [ValidatePattern('\w')]
        [string]$UserAgent = $script:DefaultUA,

        [Parameter(
            HelpMessage = 'Return the raw response returned by the API'
        )]
        [switch]
        $Raw,
        [Parameter(
            HelpMessage = 'Return the current status of rate-limit quota'
        )]
        [switch]$APIStatus
    )

    begin {

        if (-not $ApiKey) {

            # Try and retrieve API Key stored in AppData
            try {
                Write-Verbose "Attempting to retrieve API Key from encrypted file."
                
                $APIKey = Get-EmailRepAPIKey

                Write-Verbose "API Key found!"

                if ($null -eq $ApiKey) {
                    Write-Warning "API Key is not set. Please set APIKey parameter or use 'Set-EmailRep -APIKey'"
                }
            }
            catch {
                Write-Warning "API Key not set and/or could not be retrieved. Please try again..."
                Write-Error $_.Exception.Message

                break
            }
        }

        $baseUrl = "https://emailrep.io"

        switch ($ApiKey) {
            $null {
                $headers = @{
                    "Accept" = "application/json"
                }
            }
            '' {
                $headers = @{
                    "Accept" = "application/json"
                }
            }
            default {
                $headers = @{
                    "Accept" = "application/json"
                    "Key"    = $ApiKey
                }
            }
        }

    }

    process {

        foreach ($Email in $EmailAddress) {

            if ($summary) {
                $queryUrl = "{0}/{1}?summary=true" -f $baseUrl, $Email
            }
            else {
                $queryUrl = "{0}/{1}" -f $baseUrl, $Email
            }

            try {

                $r = Invoke-WebRequest -Method GET -Uri $queryUrl -Headers $headers -UserAgent $UserAgent
                $j = $r.content | ConvertFrom-Json

                switch ( $r.StatusCode ) {
                    200 { $Status = 'OK' }
                    400 { $Status = 'Bad Request: Invalid email' }
                    401 { $Status = 'Unauthorized: invalid api key (for authenticated requests)' }
                    429 { $Status = 'Too Many Requests: too many requests. contact us for an api key' }
                }

                if ($r.StatusCode -ne 200) {
                    Write-Host $Status
                    throw
                }


                if ($raw) {
                    return $j
                }
                else {
                
                    $report = [PSCustomObject]@{
                        email      = $j.email
                        reputation = $j.reputation
                        suspicious = $j.suspicious
                        references = $j.references

                    }

                    if ($Summary) {
                        $report | Add-Member -MemberType NoteProperty -Name 'summary' -Value $j.summary
                    }

                    foreach ($p in $j.details.psobject.properties) {
                        $report | Add-Member -MemberType NoteProperty -Name $p.Name -Value $p.Value
                    }

                    if ($APIStatus) {
                        if ($r.headers['X-Rate-Limit-Daily-Remaining']) {
                            Write-Host -ForegroundColor Yellow "Daily queries remaining: $($r.headers['X-Rate-Limit-Daily-Remaining'])"
                        }
                        if ($r.headers['X-Rate-Limit-Monthly-Remaining']) {
                            Write-Host -ForegroundColor Yellow "Monthly queries remaining: $($r.headers['X-Rate-Limit-Daily-Remaining'])"
                        }
                    }

                    return $report
                }
            }
            catch {
                $_.Exception
            }
        }

    }

}